### ISSUE TITLE
no bounds checking when slicing `s[start..][0..len]` arrays and array pointers in certain cases

### ISSUE THREAD

**Issue by xdBronch** [ziglang/zig#18382]
### Zig Version

0.12.0-dev.1843+fd98fc1c5

### Steps to Reproduce and Observed Behavior

```zig
test {
    var buf: [5]u8 = undefined;
    const slice = &buf;
    var a: u32 = 6;
    _ = &a;
    _ = buf[a..][0..10];
    _ = slice[a..][0..10];
}
```
this test passes even though `[a..][0..10]` is far out of bounds of `slice`/`buf`. if we instead annotate `slice` as a slice rather than array pointer
```zig
test {
    var buf: [5]u8 = undefined;
    const slice: []u8 = &buf;
    var a: u32 = 6;
    _ = &a;
    _ = buf[a..][0..10];
    _ = slice[a..][0..10];
}
```
the test now fails
if the secondary slicing is runtime known the test also fails
```zig
test {
    var buf: [5]u8 = undefined;
    const slice = &buf;
    var a: u32 = 6;
    _ = &a;
    _ = buf[a..][0..a];
    _ = slice[a..][0..a];
}
```

### Expected Behavior

safety checks should be emitted for all cases, test fail

**Comment by xdBronch** [comment]
i keep finding more quirks, if the start of the second slice is > 0 it fails
```zig
test {
    var buf: [5]u8 = undefined;
    const slice = &buf;
    var a: u32 = 6;
    _ = &a;
    _ = buf[a..][1..10];
    _ = slice[a..][1..10];
}
```
im not sure what an appropriate title would even be atp so i'll leave it

**Comment by nektro** [comment]
not confirmed but may have started since https://github.com/ziglang/zig/pull/15519

**Comment by xdBronch** [comment]
that definitely seems correct, iiuc that pr specifically special cases when the start of the second slice is 0 which would explain what i found in my comment above. i was going to build zig on the commit immediately before that pr but i dont have the right LLVM version

**Comment by xdBronch** [comment]
i was able to get a build of d71a43ec2 from the website and the first test case fails, so confirmed regression from #15519

**Comment by dweiller** [comment]
> ### Expected Behavior
> 
> safety checks should be emitted for all cases, test fail

I think the correct behaviour for the first two cases should be a compile error as the length of the by-length slice of the array is comptime known to be too large. PR incoming with this change.

**Comment by xdBronch** [comment]
thats a good point, the comment i left doing `_ = buf[a..][1..10];` should also be a compile error since `[1..10]` is OOB

**Comment by dweiller** [comment]
> thats a good point, the comment i left doing `_ = buf[a..][1..10];` should also be a compile error since `[1..10]` is OOB

With the current way things work, making that case a compile error would require a rework of slicing semantics. Currently the pattern `buf[start..][0..len]` is a special case that is a 'by-length' slice and if you want the improved comptime handling of slices with known length, but not necessarily known start you need to rewrite it as `buf[a+1..][0..9]`.

**Comment by andrewrk** [comment]
Fix reverted in bd46410419086acd274e33b47d9ae5dc1f678a1b.
